### PR TITLE
Add inline function calls as scheduler dependencies

### PR DESCRIPTION
**Pull request by awnawab**
One of the recommended ways of declaring inline function calls is via an explicit interface. This PR adds inline function calls declared via an explicit interface as scheduler dependencies. The more modern way of declaring inline function calls is to wrap them in a module and use module imports. Adding this functionality requires a larger update to the scheduler, and this PR is being prepared.

### CODE REVIEW

File: loki/transform/dependency_transform.py [hunk 1]
```diff
@@ -194,6 +219,39 @@ def rename_imports(self, source, imports, **kwargs):
...
+
+        # Transformer map to remove any outdated interfaces
+        removal_map = {}
+
+        for i in intfs:
+            for b in i.body:
+                if isinstance(b, Subroutine):
+                    target_symbol = b.name.lower()
+
+                    if targets is not None and target_symbol.upper() in targets:
```
Comment (reuterbal) [inline loki/transform/dependency_transform.py:242]: Better to use the implicit case-insensitive string comparison: `b.name in targets`

**Review by reuterbal — CHANGES_REQUESTED** [review]
_Note: This is based on #47 and #48 and should be merged after those have landed._

Very nicely done and a tremendously useful addition! Many thanks.
I have left a few comments how I think this could be improved. Most of them have to do with the fact that we are not cleanly encoding the concepts in the Scheduler at the moment and therefore there's not a good way for you to hook these changes in.
However, I would like to change this later on in a proper Scheduler overhaul and have therefore left a few proposals how to do this more aligned with the current way.

File: loki/frontend/regex.py [hunk 2]
```diff
@@ -443,7 +443,8 @@ class SubroutineFunctionPattern(Pattern):
 
     def __init__(self):
         super().__init__(
-            r'^[ \t\w()=]*?(?P<keyword>subroutine|function)[ \t]+(?P<name>\w+)\b.*?$'
+            r'^(?P<return_type>real(\(kind=.+\))?|integer(\(kind=.+\))?|logical(\(kind=.+\))?|type\(.+\))?'
```
[the remark this answers is not in the document]
Comment (awnawab) [reply loki/frontend/regex.py:446]: Actually adding a named group for return_type was overkill, I never actually used the return_type but was just trying to capture it for completeness. It turns out that the only functionality I needed was checking for the `=` sign, and you added that in this commit: https://github.com/ecmwf-ifs/loki/commit/35605560c031bbff4d88b6866798851b5bf51abc. I prefer your implementation of a catch-all `prefix` named group, which can then be split as and when needed, so I'll return that line to how it is in main. I'll also update the `test_regex_function_inline_return_type` test. Thanks!

[the remark this answers is not in the document]
Comment (awnawab) [reply loki/frontend/regex.py:446]: Actually adding a named group for return_type was overkill, I never actually used the return_type but was just trying to capture it for completeness. It turns out that the only functionality I needed was checking for the `=` sign, and you added that in this commit: https://github.com/ecmwf-ifs/loki/commit/35605560c031bbff4d88b6866798851b5bf51abc. I prefer your implementation of a catch-all `prefix` named group, which can then be split as and when needed, so I'll return that line to how it is in main. I'll also update the `test_regex_function_inline_return_type` test. Thanks!

File: loki/bulk/item.py [hunk 3]
```diff
@@ -520,6 +531,44 @@ def _variable_to_type_name(self, var):
...
+    def inline_function_interfaces(self):
+        """
+        Inline functions declared in the corresponding :any:`Subroutine`,
+        or its parent :any:`Module`, via an explicit interface.
+        """
+
+        names = ()
+        interfaces = self.routine.interfaces
+
+        if isinstance(self.scope, Module):
```
[the remark this answers is not in the document]
Comment (awnawab) [reply loki/bulk/item.py:544]: Thanks for the tip! And for the opportunity to learn a new Python operator ;-)

[the remark this answers is not in the document]
Comment (reuterbal) [reply loki/bulk/item.py]: That one is a controversial one but I consider it to be actually useful ;-)

File: loki/transform/dependency_transform.py (hunk 1 above)
Comment (reuterbal) [reply loki/transform/dependency_transform.py]: Yes, but `b.name` is a symbol that has the mixin, and it works in this direction as well:
```python
>>> l = ['a', 'b', 'c']
>>> from loki import Variable
>>> Variable(name='a') in l
True
>>> Variable(name='A') in l
True
>>> Variable(name='aa') in l
False
```

Comment (awnawab) [reply loki/transform/dependency_transform.py]: Ah yes, my bad, sorry, was suffering from a case of "ramadan brain" there 😆

File: loki/frontend/regex.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (reuterbal) [reply loki/frontend/regex.py]: Great, makes sense

File: loki/transform/dependency_transform.py [hunk 4]
```diff
@@ -94,10 +96,29 @@ def transform_subroutine(self, routine, **kwargs):
...
+        self.rename_interfaces(routine, intfs=intfs, **kwargs)
+
         if role == 'kernel' and self.mode == 'strict':
             # Re-generate C-style interface header
             self.generate_interfaces(routine)
 
+    def update_result_var(self, routine):
+        """
+        Update name of result variable for function calls.
+        """
```
Comment (reuterbal) [inline loki/transform/dependency_transform.py:110]: Spotted this only now. Does this work as expected with functions that are declared with a different result name?

```f90
function my_func result(ret)
    integer ret
    ret = 5
end function my_func
```

**Review by reuterbal — COMMENTED** [review]
Many thanks for addressing the comments. Looks great now. One more question about the dependency injection transformation but otherwise GTG!

File: loki/transform/dependency_transform.py (hunk 4 above)
Comment (awnawab) [reply loki/transform/dependency_transform.py:110]: Ah, great catch! You are right, it won't work. It needs a test to check if `result_name` matches `Subroutine.name`. I'll update and add a test 👍

**Review by reuterbal — APPROVED** [review]
Looks great, thanks for confirming and fixing!

File: tests/test_scheduler.py [hunk 5]
```diff
@@ -1558,3 +1558,31 @@ def test_scheduler_qualify_names():
...
+    my_config = config.copy()
+    my_config['routine'] = [
+        {
+            'name': 'driver',
+            'role': 'driver'
+        }
+    ]
+
+    scheduler = Scheduler(paths=here/'sources/projInlineCalls', config=my_config, frontend=frontend)
+    print(scheduler.items)
```
Comment (mlange05) [inline tests/test_scheduler.py:1578]: Debug leftover?

**Review by mlange05 — COMMENTED** [review]
Looks very good, and again, many thanks for increased test coverage. There's still one corner case flagged by @reuterbal that needs addressing (unless I missed something?), and I'highlighted a (presumably) debug print that we should probably drop. Otherwise GTG after these are addressd.

File: tests/test_scheduler.py (hunk 5 above)
Comment (awnawab) [reply tests/test_scheduler.py:1578]: Ah yes that is a debug print, I'll remove that and push an update. And are you referring to B's spot that `update_result_var` won't work for a function with an explicitly declared result name? If so, I've fixed that one.

File: loki/transform/dependency_transform.py (hunk 4 above)
Comment (mlange05) [reply loki/transform/dependency_transform.py:111]: Sorry, did not understand the fix, but cleared up offline. All good now

File: tests/test_scheduler.py (hunk 5 above)
Comment (mlange05) [reply tests/test_scheduler.py]: Yes, sorry had not understood the fix, but happy now!

**Review by mlange05 — APPROVED** [review]
OK, all addressed, GTG from me.
